src/Management/src/Endpoint/Actuators/HttpExchanges/HttpExchangesDiagnosticObserver.cs

```diff
@@ -0,0 +1,216 @@
+
+        if (!options.IncludePathInfo)
+        {
+            filteredUri.Path = string.Empty;
+        }
+
+        if (!options.IncludeQueryString)
+        {
+            filteredUri.Query = string.Empty;
+        }
+
+        return filteredUri.Uri;
+    }
+
+    private HttpExchange GetHttpExchange(HttpContext context, TimeSpan duration)
+    {
+        HttpExchangesEndpointOptions options = _optionsMonitor.CurrentValue;
+
+        Uri requestUri = GetRequestUri(context.Request);
+        Dictionary<string, StringValues> requestHeaders = GetHeaders(context.Request.Headers);
+
+        var request = new HttpExchangeRequest(context.Request.Method, requestUri, requestHeaders, context.Connection.RemoteIpAddress?.ToString());
+
+        Dictionary<string, StringValues> responseHeaders = GetHeaders(context.Response.Headers);
+
+        var response = new HttpExchangeResponse(context.Response.StatusCode, responseHeaders);
+
+        HttpExchangePrincipal? principal = GetUserPrincipal(context);
+
+        string? sessionId = null;
+
+        if (options.IncludeSessionId)
+        {
+            sessionId = GetSessionId(context);
+        }
+
+        HttpExchangeSession? session = sessionId == null ? null : new HttpExchangeSession(sessionId);
+
+        return new HttpExchange(request, response, DateTime.UtcNow, principal, session, duration);
+    }
+
+    private static Dictionary<string, StringValues> GetHeaders(IHeaderDictionary headers)
+    {
+        var result = new Dictionary<string, StringValues>();
+
+        foreach ((string key, StringValues value) in headers)
+        {
+            result.Add(key, value);
+        }
+
+        return result;
+    }
+
+    private static Dictionary<string, StringValues> FilterHeaders(IDictionary<string, StringValues> headers, IList<string> allowedHeaders)
+    {
+        var filteredHeaders = new Dictionary<string, StringValues>();
+
+        foreach ((string key, StringValues value) in headers)
+        {
+            if (HeaderShouldBeRedacted(key, allowedHeaders))
+            {
+                filteredHeaders[key] = Redacted;
+            }
+            else
+            {
+                filteredHeaders.Add(key, value);
+            }
+        }
+
+        return filteredHeaders;
+    }
+
+    private static bool HeaderShouldBeRedacted(string currentHeader, IList<string> allowedHeaders)
+    {
+        return allowedHeaders is not [""] && !allowedHeaders.Any(header => header.Equals(currentHeader, StringComparison.OrdinalIgnoreCase));
+    }
+
+    internal static Uri GetRequestUri(HttpRequest request)
+    {
+        ArgumentNullException.ThrowIfNull(request);
+
+        var uriBuilder = new StringBuilder();
+        uriBuilder.Append(request.Scheme);
+        uriBuilder.Append(Uri.SchemeDelimiter);
+        uriBuilder.Append(request.Host);
+        uriBuilder.Append(request.Path);
+        uriBuilder.Append(request.QueryString);
+
+        return new Uri(uriBuilder.ToString());
+    }
```

Comment:
I find this rather brittle. For example, does `.QueryString` include the `?` or not? Why not just use `request.GetEncodedUrl()`? (looking into its implementation, it's more efficient too)